### PR TITLE
feat: Add support for SR-IOV Network Adapters

### LINKED ISSUE #1136: Add support for SR-IOV network adapters
### Description

When deploying a VM to a particular host you may want to attach a network adapter directly to an SR-IOV enabled physical NIC (bypassing a vSwitch)

To do this in the GUI, you simply attach a network adapter as normal and select 'SRIOV' instead of 'vmxnet3' (the default) and provide a link to the physical NIC. You still need to provide the name of a virtual network (from which you inherit various VLAN settings)

image

### Potential Terraform Configuration

```hcl
resource "vsphere_virtual_machine" "vm" {
   # ... other configuration ...
  host_system_id      = data.vsphere_host.host.id
  memory              = var.memory
  memory_reservation  = var.memory # Need to reserve memory for SR-IOV to work
  network_interface  {
    network_id        = data.vsphere_network.network.id
    adapter_type      = sriov
    physical_function = "0000:3b:00.1" 
  }
  ... other network_interfaces... 
}
```

Although there may be a natural extension to make a data type representing the physical NIC and the physical PCI is read off that data type (I think we can build off @bill-rich 's work in https://github.com/hashicorp/terraform-provider-vsphere/pull/1099

### References

Th MOP to do this is [here](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.networking.doc/GUID-EE03DC6F-32CA-42EF-98FC-12FDE06C0BE0.html
)

This MOP is in two parts:

1. enable SR-IOV on the host:  [this MOP](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.networking.doc/GUID-BEA364D2-7A9D-44E6-BA2D-92A07516B6B9.html)
2. [Assign a Virtual Function as SR-IOV Passthrough Adapter to a Virtual Machine](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.networking.doc/GUID-898A3D66-9415-4854-8413-B40F2CB6FF8D.html)

This work essentially adds support for the second part of the MOP in terraform. 

I've started work on this enhancement, and have a PR that does deploy the VM correctly but is hitting an issue with UnitNumber ordering [see this issue](https://github.com/vmware/govmomi/issues/2060)

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by sunnycarter**
### Description

This pull request adds SR-IOV network_interface function to terraform, so that you can add SR-IOV Nics to a VM.

SR-IOV, Single Root I/O Virtualization, allows a single NIC (termed the Physical Function - PF) to be shared between a bounded number of VMs, providing a Virtual Function (VF) network interface to each VM. It is used for high-performance throughput on the network interface.

An SR-IOV network interface subresource would be declared in terraform main.tf as follows:

```hcl
resource "vsphere_virtual_machine" "vm" {
  # ... other configuration ...
  host_system_id      = data.vsphere_host.host.id
  memory              = var.memory
  memory_reservation  = var.memory
  network_interface  {
    network_id        = data.vsphere_network.network.id
    adapter_type      = sriov
    physical_function = "0000:3b:00.1" 
  }
  ... other network_interfaces... 
}
```

If you set the adapter_type to sriov, you must also do the following:

* ensure the host_system_id is set
* ensure that memory_reservation == memory
* set the physical_function on the interface to a SRIOV enabled physical adapter address 
* Additionally, bandwidth settings are not relevant to SRIOV nics.

There are limitations to the use of SRIOV NICS, due to fiddlyness with their PCI Bus unit numbers. (Non-SRIOV interfaces use unit numbers 7-17, whereas SRIOV interfaces use unit numbers 45-36 decreasing.) As much of the code for network adapters uses unit numbers as akin to index number of the interface in main.tf, this has made the changes fiddly. The limitations created here mean: 

-  All SRIOV Nics must be declared after all non-SRIOV Nics. 
-  You cannot add or delete non-SRIOV nics and update, if you have any SRIOV nics in the config.

### Acceptance tests
- [ ] Have you added an acceptance test for the functionality being added?  NO
- [ ] Have you run the acceptance tests on this branch? NO

I have not run acceptance tests as my vsphere environment does not have access to any NAS storage, so I cannot set environment variable TF_VAR_VSPHERE_NAS_HOST or TF_VAR_VSPHERE_NFS_PATH, so tests are failing with ```
make testacc TESTARGS="-run=TestAccResourceVSphereVirtualMachine_basic"
<snip>
Error: error mounting datastore: host "<TF_VAR_VSPHERE_ESXI2>": ServerFaultCode: A specified parameter was not correct:

          with vsphere_nas_datastore.ds1,
          on terraform_plugin_test.tf line 29, in resource "vsphere_nas_datastore" "ds1":
          29: resource "vsphere_nas_datastore" "ds1" {```
. I'd appreciate guidance on how to get around this, and which acceptance tests to run if it is possible. I've made notes on manual testing below.

### Manual testing
The following manual tests have been done:
Error cases:
- Missing host
- Invalid Physical function
- Memory not reserved
- SRIOV not enabled on physical function
- Simple - change adapter type to sriov from vmxnet
- Simple - change adapter type to vmxnet to sriov
- Missing physical function on sriov nic
- Physical function on vmxnet nic
- Wrong host

Apply new:
- 2x VMXnet then 2xSRIOV
- Vmxnet, SRIOV, vmxnet, SRIOV
- 10 VMXnet, 10 SRIOV
- 10 VMXnet, 11 SRIOV
- 2 SRIOV
- 2xVM pair with 2VMxnet 2SRIOV
- Bandwidth limit on vmxnet
- e1000 network
- 3 non-SRIOV

Update (new interfaces):
- Additional SRIOV new resource indexes
- Additional VMXnet new resource indexes no SRIOV
- 2VMXnet, 2 SRIOV add new VMXnet before SRIOV
- Switch physical adapter on two SRIOV NICs
- Vmxnet bandwidth limit

Delete interfaces:
- Delete SRIOV NICs from end of list
- Delete SRIOV NICs within list
- Delete Vmxnet NIC at end of VMxnet list with no SRIOV
- Delete Vmxnet within list containing SRIOV after
- Delete vmxnet within list containing only Vmxnet

Import:
- 2x VMXnet then 2xSRIOV
- 6 VMXnet, 6 SRIOV
- e1000
- Bandwidth limit non default

Bandwidth stuff:

- One vmxnet without, one with non-defaults, one sriov, apply
- same as above but import then plan
- Add new SRIOV NIC - Update
- Update vmxnet bandwidth setting
- Import after the above
- Bandwidth setting on SRIOV nic

Non-interface tests:

- Update some other non NIC function
- Import without SRIOV

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vsphere/blob/master/CHANGELOG.md):

```release-note
FEATURES:

resource/vm/network_interface: Add adapter_type sriov which requires the network_interface to have a physical_function declared.  This new network_interface adapter_type allows vsphere_virtual_machine to be created with SRIOV Network Adapters. See PR for more details. (@@@TBD PR number)
```
### References

https://github.com/vmware/govmomi/issues/2060 
Resolves #1136

### CODE REVIEW

**Comment by andrew-lee-metaswitch** [comment]
This fixes https://github.com/hashicorp/terraform-provider-vsphere/pull/1417

**Comment by andrew-lee-metaswitch** [comment]
Any update on this @koikonom  and other Authorized users /approvers? What's the process for getting this reviewed and merged?

**Comment by andrew-lee-metaswitch** [comment]
Any update on this? is this ever going to be reviewed / merged?

**Comment by sunnycarter** [comment]
Hi there - is there any update on the review for this?  What are the next steps please?

**Comment by tenthirtyam** [comment]
@sunnycarter @andrew-lee-metaswitch  - could you review and address the merge conflicts? If so, I’ll take a look at reviewing and onboarding this capability.

Meanwhile, I’ll mark this as draft.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
@sunnycarter @andrew-lee-metaswitch - could you review and address the merge conflicts? If so, I’ll take a look at reviewing and onboarding this capability.

Ryan Johnson
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Review by tenthirtyam — DISMISSED** [review]
@sunnycarter @andrew-lee-metaswitch - could you review and address the merge conflicts? If so, I’ll take a look at reviewing and onboarding this capability.

Ryan Johnson
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by sunnycarter** [comment]
Yes I will look into merge conflicts next week if possible. It's been a long time so I think it may take me a while to work through them. Thanks for saying you'll then look into it - we are very keen for this to get reviewed and merged in still.

**Comment by sunnycarter** [comment]
@tenthirtyam please can I ask for some guidance on sorting out these merge conflicts?
When I wrote this Pull Request, the vmware/govmomi code was included within this terraform-provider-vsphere repo, and I had to make some changes to one file (which was `vendor/github.com/vmware/govmomi/object/virtual_device_list.go`, now deleted from this repo)

The vmware/govmomi code source no longer seems to be in this repo is now referenced to the source repo for govmomi, and the file that I want to change now resides here: https://github.com/vmware/govmomi/blob/master/object/virtual_device_list.go#L921   (I can't find the changelog that made this change) 

I assume therefore I would need to submit a PullRequest on that repo, and get the version bumped, then pull in that new version of govmomi into this PullRequest before proceeding?

Am I correct, and do you know anything about the speed of that process?

**Comment by sunnycarter** [comment]
@tenthirtyam Ryan Johnson,
I've made an attempt at dealing with the merge conflicts here, which included creating a Pull Request on govmomi : https://github.com/vmware/govmomi/pull/2957 

My question above still stands as to whether this was the correct course of action, as I'm not quite sure what the history of this file was. The earlier commits to this terraform pull request show the previous change I made in this repo.  Annoyingly, I can't remember if the change to that file was purely cosmetic, or was actually required for these SRIOV changes to work. I'm going with the assumption that it was required for now.

**Comment by tenthirtyam** [comment]
Hi Sunny - I think you make the best steps forward here. The provider no longer requires the vendor directory and instead references this from the upstream.

 I've cc: the maintainer in the upstream `vmware/govmomi` issue and have added an `upstream/govmoni` label to this PR since we'd need to await an upstream merge. Afterwhich, we'll open a separate PR to update `vmware/govmomi` for this provider.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
Upstream [PR](https://github.com/vmware/govmomi/pull/2957) merged.

**Comment by sunnycarter** [comment]
@tenthirtyam Hi Ryan,
Now that the Upstream PR https://github.com/vmware/govmomi/pull/2957 merged what are the next steps to getting this PR merged please?    Is there a way you (or I) can request a new release of govmomi so that it can be pulled into this PR and we can get it merged??

**Comment by tenthirtyam** [comment]
Yes, we'd need an upstream release of `vmware/govmomi` and subsequent bump (and testing) here.

**Comment by tenthirtyam** [comment]
Good news is that @akutz just released [`vmware/govmomi@v0.30.0`](https://github.com/vmware/govmomi/releases/tag/v0.30.0) today.

**Comment by tenthirtyam** [comment]
Dependent on https://github.com/hashicorp/terraform-provider-vsphere/pull/1801.

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource_test.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource_test.go:1]: QQ - why was the test deleted vs extended?

File: go.sum
Comment (tenthirtyam) [inline go.sum:9]: I don't think we need to make changes here in `go.sum` - that would be picked up by the release process.

File: website/docs/r/virtual_machine.html.markdown
Comment (tenthirtyam) [inline website/docs/r/virtual_machine.html.markdown:969]:
```suggestion
#### Using SR-IOV Network Interfaces
```

File: website/docs/r/virtual_machine.html.markdown
Comment (tenthirtyam) [inline website/docs/r/virtual_machine.html.markdown:992]:
```suggestion
* Adding, modifying, and deleting SR-IOV NICs is supported but requires a VM restart.
```

File: website/docs/r/virtual_machine.html.markdown
Comment (tenthirtyam) [inline website/docs/r/virtual_machine.html.markdown:994]:
```suggestion
* Modifying the number of non-SR-IOV (_e.g._, VMXNET3) interfaces when there are SR-IOV interfaces existing is
```

File: website/docs/r/virtual_machine.html.markdown
Comment (tenthirtyam) [inline website/docs/r/virtual_machine.html.markdown:996]:
```suggestion
  non-SR-IOV to SR-IOV or vice-versa). To work around this, delete all SRIOV NICs for one Terraform apply, and re-add 
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:496]:
```suggestion
			return fmt.Errorf("trying to use an SR-IOV network interface but target host is not known")
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:516]:
```suggestion
				return fmt.Errorf("unable to find SR-IOV physical adapter %s on host %s", sriovPhysicalAdapter, host.Name())
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:534]:
```suggestion
				return fmt.Errorf("unable to find SR-IOV physical adapter PCI passthrough Id %s on host %s", sriovPhysicalAdapters[adapterIdx], host.Name())
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:568]:
```suggestion
			return fmt.Errorf("trying to use SR-IOV NIC but memory reservation is not equal to memory, set memory_reservation equal to memory on VM")
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:1322]:
```suggestion
			return fmt.Errorf("changing the network_interface list such that there is a change in adapter_type to"+
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:1341]:
```suggestion
			return fmt.Errorf("invalid bandwidth properties on sriov network interface. " +
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:1340]: "SUNNY"

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:1357]:
```suggestion
	// 6.0 and higher. They are not relevant for SR-IOV networks in either case.
```

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
Comment (tenthirtyam) [inline vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:1466]:
```suggestion
			return fmt.Errorf("all ten SR-IOV device units are currently in use on the PCI bus. cannot assign SR-IOV network.")
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Did a quick review and spotted some minot correction for terms and error msgs.

File: go.sum
Comment (sunnycarter) [reply go.sum]: I've reverted that file to origin/master

File: vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource_test.go
Comment (sunnycarter) [reply vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource_test.go:1]: The unit test was testing a single function which I have removed.  I replaced the logic from nicUnitRange with something that in my mind was simpler, and works for both the non-sriov and sriov cases, where unit numbers either go from 7 descending or 45 descending.  Instead of using nicUnitRange to determine the length of array that needed to be populated, I created arrays of standard length, populate them and then remove any gaps, which simplifies the dealing with non-sriov/sriov cases.  
Example here: https://github.com/hashicorp/terraform-provider-vsphere/pull/1417/files#diff-e8de2d20cbf49f2db545eafc50b5230673e69274016a107bc7c75f6adcb1d635R594

As there was only one unit test and it was for a function that's gone, I removed the test too.

**Comment by sunnycarter** [comment]
Thank you very much for this review.  I've resolved all the requested changes apart from the one that was a question, where I've tried to explain my reasoning. 

I'm handing back to you in case you wanted to review further. I know we are waiting for https://github.com/hashicorp/terraform-provider-vsphere/pull/1801 to be merged so that it can be pulled into this one before this can be merged?

**Comment by tenthirtyam** [comment]
Added to v2.3.0 queue for testing based on status of reviews to date - could push to v2.4.0.

**Comment by JoshuaFurman** [comment]
If possible would love to push for this to be added to v2.3.0 release as this is a feature my team and I could very much use

**Comment by tenthirtyam** [comment]
> If possible would love to push for this to be added to v2.3.0 release as this is a feature my team and I could very much use

Please use the recommended community practice  of adding a 👍 reaction to the linked enhancement issue.

**Comment by tenthirtyam** [comment]
@sunnycarter - I noticed that one of the test is failing:

```
==> Checking that code complies with gofmt requirements...
[5](https://github.com/hashicorp/terraform-provider-vsphere/actions/runs/3733914469/jobs/6336346585#step:6:6)
gofmt needs running on the following files:
[6](https://github.com/hashicorp/terraform-provider-vsphere/actions/runs/3733914469/jobs/6336346585#step:6:7)
./vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
[7](https://github.com/hashicorp/terraform-provider-vsphere/actions/runs/3733914469/jobs/6336346585#step:6:8)
You can use the command: `make fmt` to reformat code.
[8](https://github.com/hashicorp/terraform-provider-vsphere/actions/runs/3733914469/jobs/6336346585#step:6:9)
make: *** [GNUmakefile:46: fmtcheck] Error 1
[9](https://github.com/hashicorp/terraform-provider-vsphere/actions/runs/3733914469/jobs/6336346585#step:6:10)
Error: Process completed with exit code 2
```

**Comment by JoshuaFurman** [comment]
Can I inquire about the status of this PR? This feature would prove extremely useful for my team

**Comment by tenthirtyam** [comment]
> Can I inquire about the status of this PR? This feature would prove extremely useful for my team

Please refer to the linked milestone.

**Comment by tenthirtyam** [comment]
> Sorry about that, I've run it now

Hi @sunnycarter - there still appears to be a conflict on `vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource_test.go`.

**Comment by appilon** [comment]
Hello Folks!

We will make an effort to prioritize contributions, which we sincerely appreciate. The provider is long overdue for some technical debt work. This feature along with a few others, are innately harder to test. The current focus is to streamline and improve the testing and CI for this provider, [this PR](https://github.com/hashicorp/terraform-provider-vsphere/pull/1847) was a small step in that direction, there should be more PRs soon around this.

Once we can establish some CI workflow for the provider, we will make an effort to prioritize feature work. We appreciate your patience and contributions.

**Comment by JoshuaFurman** [comment]
> 

Thank you for working towards this!

**Comment by holmesb** [comment]
This one still on track for v2.4 @appilon? I note CI PR #1847 was merged in Feb. Many thanks.

**Comment by tenthirtyam** [comment]
> This one still on track for v2.4 @appilon? I note CI PR #1847 was merged in Feb. Many thanks.

No, we're planning to cut a v2.4.0 release earlier and push out the remaining to a later release.

**Comment by holmesb** [comment]
Great work on this PR @sunnycarter. I've been doing some testing. Plan fails when VM's `host_system_id` or it's SR-IOV NIC `physical_function` are only know after apply, eg because they're obtained from a datasource or another resource. I'd expect plan to show "(known after apply)" and continue normally, but instead it errors: "trying to use an SR-IOV network interface but target host is not known" or "physical_function must be set on SR-IOV Network interface". 

Workaround is to replace [NetworkInterfaceDiffOperation](https://github.com/sunnycarter/terraform-provider-vsphere/blob/de69c0818af3e016418dd8edb9e9df6cf324f547/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go#L419) and [ValidateDiff ](https://github.com/sunnycarter/terraform-provider-vsphere/blob/de69c0818af3e016418dd8edb9e9df6cf324f547/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go#L1355)functions in this PR with original develop branch's. Works, but your helpful validation and descriptive error messages are lost.

**Comment by holmesb** [comment]
Workaround is in my [repo](https://github.com/holmesb/terraform-provider-vsphere/tree/patch_pr_1417).

**Comment by holmesb** [comment]
@andrew-lee-metaswitch, you're right, there's no [resource](https://github.com/hashicorp/terraform-provider-vsphere/issues/1685) or datasource for retrieving a PCI address to feed into `physical_function`. We've implemented our own workaround using remote_exec. Not ideal, but it works. Whatever fix you're thinking of for `host_system_id` should also apply to `physical_function`, so when #1685 lands, these known-at-apply PCI addresses won't break SR-IOV NICs.

**Pull request closed without merging**
